### PR TITLE
 ENH: Add equals_nans kwarg to np.unique

### LINKED ISSUE #20326: Revert returning a single NaN for `np.unique` in 1.22.0?
[This mailing list thread](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/DBJZNCNBXTNU2EWKDZMP7MDQLOC46GNV/?sort=thread) discusses reverting the change in `1.21.0` to make `unique` return a single `nan` in its output when the input array contains multiple `nan`s. This issue is to track that, needs a decision before `1.22.0rc1` I think.

### PR DESCRIPTION
**Pull request by rjeb**
Addresses https://github.com/numpy/numpy/issues/20326

np.unique previously had it's functionality changed so NaN values would be treated as non-unique.
This PR puts the functionality into the kwarg equal_nans(default: True).

### CODE REVIEW

**Comment by rjeb** [comment]
@mattip I have some changes from something I was trying before that I think addresses #20326.

File: doc/release/upcoming_changes/21623.new_feature.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+New parameter ``equal_nans`` added to `np.unique`
+-----------------------------------------------------------------------------------
+
+`np.unique` was previously changed to treat NaN values as equal. Now this functionality is decided by setting the ``equal_nans`` kwarg to True or False. True is the default behavior.
```
Comment (mattip) [inline doc/release/upcoming_changes/21623.new_feature.rst:4]: Rephrase and reflow to 80 columns
```suggestion
`np.unique` was changed in 1.21 to treat all ``NaN`` values as equal and return
a single ``NaN``. Setting ``equal_nans=False`` will restore pre-1.21 behavior
to treat ``NaNs`` as unique. Defaults to ``True``.

File: numpy/lib/arraysetops.py [hunk 2]
```diff
@@ -162,8 +162,10 @@ def unique(ar, return_index=False, return_inverse=False,
     return_counts : bool, optional
         If True, also return the number of times each unique item appears
         in `ar`.
+    equal_nans : bool, optional
+        If True, collapses multiple NaN values in return array into 1
 
-        .. versionadded:: 1.9.0
+        .. versionchanged: NumPy 1.24
```
Comment (mattip) [inline numpy/lib/arraysetops.py:168]:
```suggestion
        .. versionchanged: 1.24
```

**Comment by mattip** [comment]
@asmeurer is this sufficient to bring np.unique(a, `equal_nans=False`) into compliance with the array API? 

Bikeshedding on `equal_nans`, should it be maybe `collapse_nans` instead? They are not really "equal". What do others think?

File: numpy/lib/arraysetops.py [hunk 3]
```diff
@@ -334,7 +337,8 @@ def _unique1d(ar, return_index=False, return_inverse=False,
         aux = ar
     mask = np.empty(aux.shape, dtype=np.bool_)
     mask[:1] = True
-    if aux.shape[0] > 0 and aux.dtype.kind in "cfmM" and np.isnan(aux[-1]):
+    if (equal_nans and aux.shape[0] > 0 and aux.dtype.kind in "cfmM" and 
+    np.isnan(aux[-1])):
```
Comment (mattip) [inline numpy/lib/arraysetops.py:341]: This always looks weird: either it lines up with the line below, or is over-indented. I forget what black does here...
```suggestion
            np.isnan(aux[-1])):
```

**Comment by mattip** [comment]
> Bikeshedding on equal_nans ...

This got a green light in the developer meeeting, so LGTM.

**Comment by seberg** [comment]
Actually, a little oops there.  My/our argument was that we already have `equal_nans`, but that is not true.  The parameter is called `equal_nan` on the other functions, so maybe that would be better?

**Comment by asmeurer** [comment]
The array API specifies that nans are not equal. See https://data-apis.org/array-api/latest/API_specification/generated/signatures.set_functions.unique_all.html#signatures.set_functions.unique_all for instance. 

We would need to invert this flag in the array_api. The unique functions in the array API have new names so it wouldn't be backwards incompatible to change the default for them.

**Comment by mattip** [comment]
@asmeurer NumPy has chosen in `np.unique` to not follow the array API, rather to reduce all nans to a single value. This PR is to allow `np.array_api.unique_all` to use `np.unique(..., equal_nan=False)` under the hood. My question is if this PR is sufficient for the implementation of `np.array_api.unique_all`, which was the original request in #20326

**Comment by asmeurer** [comment]
Yes, I believe it should be. We just need to add equal_nans=False to the wrappers in the array API.

**Pull request merged**